pybart/new_matcher.py

```diff
@@ -0,0 +1,330 @@
+                        # keep the filtered assignment for further merging
+                        edge_assignments[i].append(assignment)
+        
+        merges = [{}]
+        # for each list of possible assignments of an edge
+        for i in range(len(self.constraint.edges)):
+            new_merges = []
+            # for each merged assignment
+            for merged in merges:
+                # for each possible assignment in the current list
+                for assignment in edge_assignments[i]:
+                    # try to merge (see that there is no contradiction on hashing)
+                    just_merged = self._try_merge(merged, assignment)
+                    if just_merged:
+                        new_merges.append(just_merged)
+            if not new_merges:
+                return []
+            merges = new_merges
+        
+        for merged_assignment in merges:
+            # TODO - compare the speed of non-edge filtering here to the on-going edge filter (see previous todo)
+            if self._filter(merged_assignment, sentence):
+                # keep only required captures
+                _ = [merged_assignment.pop(name) for name in self.dont_capture_names]
+                captured_labels = {(v1, v2): labels for (k1, v1, k2, v2), labels in self.captured_labels.items()
+                                   if merged_assignment[k1] == v1 and merged_assignment[k2] == v2}
+                # append assignment to output
+                sentence_matches.append(SentenceMatch(merged_assignment, captured_labels))
+        return sentence_matches
+       
+
+class TokenMatcher:
+    # convert the constraint to a spacy pattern
+    @staticmethod
+    def _convert_to_matchable(constraints: Sequence[Token]) -> List[Tuple[str, bool, Any]]:  # TODO - change Any to whatever type we need
+        patterns = []
+        for constraint in constraints:
+            pattern = dict()
+            for spec in constraint.spec:
+                # TODO - the list is comprised of either strings or regexes - so spacy's 'IN' or 'REGEX' alone is not enough
+                #   (we need a combination or another solution)
+                if spec.field == FieldNames.WORD:
+                    # TODO - add separation between default LOWER and non default TEXT
+                    pattern["TEXT"] = {"IN", spec.value}
+                elif spec.field == FieldNames.LEMMA:
+                    pattern["LEMMA"] = {"IN", spec.value}
+                elif spec.field == FieldNames.TAG:
+                    pattern["TAG"] = {"IN", spec.value}
+                elif spec.field == FieldNames.ENTITY:
+                    pattern["ENT_TYPE"] = {"IN", spec.value}
+            
+            patterns.append((constraint.id, not constraint.optional, pattern))
+        return patterns
+    
+    def __init__(self, constraints: Sequence[Token]):
+        self.nlp = spacy.load("en_ud_model_lg")
```

Comment:
this looks wrong to me, you shouldn't create a spacy model inside the matcher.
if you need the vocab, then someone should pass you either the nlp object, or the vocab object.